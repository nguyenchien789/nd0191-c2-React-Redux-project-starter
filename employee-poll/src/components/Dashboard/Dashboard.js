import React from "react";
import { useEffect, useState } from "react";
import { Switch, List } from "antd";
import { Link } from "react-router-dom";
import { connect } from "react-redux";

const Dashboard = ({ array, answer }) => {
  const [isAnswered, setIsAnswered] = useState(false);
  const [data, setData] = useState([]);
  useEffect(() => {
    let data;
    if (isAnswered) {
      data = array.filter((item) => answer.includes(item.id));
    } else {
      data = array.filter((item) => !answer.includes(item.id));
    }
    setData(data);
  }, [isAnswered]);

  const onChange = (checked) => {
    setIsAnswered(checked);
  };

  return (
    <>
      <div>
        <Switch style={{ marginRight: "20px" }} onChange={onChange} />
        {isAnswered ? "Answered Questions" : "Unanswered Questions"}
      </div>
      <List
        itemLayout="horizontal"
        dataSource={data}
        renderItem={(item, index) => (
          <List.Item>
            <List.Item.Meta
              title={
                <Link to={`/question/${item.id}`}>
                  Would you rather {item.optionOne.text} or{" "}
                  {item.optionTwo.text}
                </Link>
              }
              description={item.author}
            />
          </List.Item>
        )}
      />
    </>
  );
};

const mapStateToProps = ({ isLogged, users, questions }) => {
  const user = users[isLogged.id];
  const array = Object.keys(questions)
    .map((key) => questions[key])
    .sort((a, b) => b.timestamp - a.timestamp);
  const answer = Object.keys(user.answers).map((key) => key);

  console.log(isLogged);
  console.log(users);
  console.log(questions);
  return {
    array: array,
    answer: answer,
  };
};

export default connect(mapStateToProps)(Dashboard);
