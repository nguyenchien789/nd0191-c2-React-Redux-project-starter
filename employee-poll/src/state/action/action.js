
export const getUsersData = (data) => {
  return {
      type: "GET_USERS",
      data
  }
}

export const getQuestionsData = (data) => {
  return {
      type: "GET_QUESTIONS",
      data
  }
}

export const userLogin = (user) => {
  return {
      type: "LOGIN",
      user
  }
}

export const userLogout = () => {
  return {
      type: "LOGOUT"
  }
}

export const addQuestion = (data) => {
  return {
      type: "ADD_QUESTION",
      data
  }
}


export const updateQuestions = (data) => {
  return {
      type: "UPDATE_QUESTIONS",
      data
  }
}

export const updateUsers = (data) => {
  return {
      type: "UPDATE_USERS",
      data
  }
}