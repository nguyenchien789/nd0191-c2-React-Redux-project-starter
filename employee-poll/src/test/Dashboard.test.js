import React from "react";
import { render, cleanup } from "@testing-library/react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import { store } from "../state/store/store";
import { _getQuestions, _getUsers } from "../_DATA";
import { getQuestionsData, getUsersData, userLogin } from "../state/action/action";
import Dashboard from "../components/Dashboard/Dashboard";

describe("Dashboard Component", () => {
  beforeAll(() => {
    Object.defineProperty(window, "matchMedia", {
      writable: true,
      value: jest.fn().mockImplementation((query) => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(), // Deprecated
        removeListener: jest.fn(), // Deprecated
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
      })),
    });
    delete window.matchMedia;
    window.matchMedia = (query) => ({
      matches: false,
      media: query,
      onchange: null,
      addListener: jest.fn(), // deprecated
      removeListener: jest.fn(), // deprecated
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
    });
  });
  afterEach(cleanup);
  it("Return a snapshot", async () => {
    const [users, questions] = await Promise.all([
      _getUsers(),
      _getQuestions(),
    ]);
    store.dispatch(getUsersData(users));
    store.dispatch(getQuestionsData(questions));
    store.dispatch(userLogin(store.getState().users["sarahedo"]));

    const component = render(
      <Provider store={store}>
        <BrowserRouter>
          <Dashboard />
        </BrowserRouter>
      </Provider>
    );
    expect(component).toBeDefined();
    expect(component).toMatchSnapshot();
  });
});
