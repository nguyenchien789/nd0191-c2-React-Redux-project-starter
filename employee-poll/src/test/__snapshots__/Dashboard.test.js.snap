// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard Component Return a snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <button
          aria-checked="false"
          class="ant-switch css-dev-only-do-not-override-1r287do"
          role="switch"
          style="margin-right: 20px;"
          type="button"
        >
          <div
            class="ant-switch-handle"
          />
          <span
            class="ant-switch-inner"
          >
            <span
              class="ant-switch-inner-checked"
            />
            <span
              class="ant-switch-inner-unchecked"
            />
          </span>
        </button>
        Unanswered Questions
      </div>
      <div
        class="ant-list ant-list-split css-dev-only-do-not-override-1r287do"
      >
        <div
          class="ant-spin-nested-loading css-dev-only-do-not-override-1r287do"
        >
          <div
            class="ant-spin-container"
          >
            <ul
              class="ant-list-items"
            >
              <li
                class="ant-list-item"
              >
                <div
                  class="ant-list-item-meta"
                >
                  <div
                    class="ant-list-item-meta-content"
                  >
                    <h4
                      class="ant-list-item-meta-title"
                    >
                      <a
                        href="/question/xj352vofupe1dqz9emx13r"
                      >
                        Would you rather 
                        deploy to production once every two weeks
                         or
                         
                        deploy to production once every month
                      </a>
                    </h4>
                    <div
                      class="ant-list-item-meta-description"
                    >
                      mtsamis
                    </div>
                  </div>
                </div>
              </li>
              <li
                class="ant-list-item"
              >
                <div
                  class="ant-list-item-meta"
                >
                  <div
                    class="ant-list-item-meta-content"
                  >
                    <h4
                      class="ant-list-item-meta-title"
                    >
                      <a
                        href="/question/vthrdm985a262al8qx3do"
                      >
                        Would you rather 
                        take a course on ReactJS
                         or
                         
                        take a course on unit testing with Jest
                      </a>
                    </h4>
                    <div
                      class="ant-list-item-meta-description"
                    >
                      tylermcginnis
                    </div>
                  </div>
                </div>
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <button
        aria-checked="false"
        class="ant-switch css-dev-only-do-not-override-1r287do"
        role="switch"
        style="margin-right: 20px;"
        type="button"
      >
        <div
          class="ant-switch-handle"
        />
        <span
          class="ant-switch-inner"
        >
          <span
            class="ant-switch-inner-checked"
          />
          <span
            class="ant-switch-inner-unchecked"
          />
        </span>
      </button>
      Unanswered Questions
    </div>
    <div
      class="ant-list ant-list-split css-dev-only-do-not-override-1r287do"
    >
      <div
        class="ant-spin-nested-loading css-dev-only-do-not-override-1r287do"
      >
        <div
          class="ant-spin-container"
        >
          <ul
            class="ant-list-items"
          >
            <li
              class="ant-list-item"
            >
              <div
                class="ant-list-item-meta"
              >
                <div
                  class="ant-list-item-meta-content"
                >
                  <h4
                    class="ant-list-item-meta-title"
                  >
                    <a
                      href="/question/xj352vofupe1dqz9emx13r"
                    >
                      Would you rather 
                      deploy to production once every two weeks
                       or
                       
                      deploy to production once every month
                    </a>
                  </h4>
                  <div
                    class="ant-list-item-meta-description"
                  >
                    mtsamis
                  </div>
                </div>
              </div>
            </li>
            <li
              class="ant-list-item"
            >
              <div
                class="ant-list-item-meta"
              >
                <div
                  class="ant-list-item-meta-content"
                >
                  <h4
                    class="ant-list-item-meta-title"
                  >
                    <a
                      href="/question/vthrdm985a262al8qx3do"
                    >
                      Would you rather 
                      take a course on ReactJS
                       or
                       
                      take a course on unit testing with Jest
                    </a>
                  </h4>
                  <div
                    class="ant-list-item-meta-description"
                  >
                    tylermcginnis
                  </div>
                </div>
              </div>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Dashboard Component Return a snapshot 2`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <button
          aria-checked="false"
          class="ant-switch css-dev-only-do-not-override-1r287do"
          role="switch"
          style="margin-right: 20px;"
          type="button"
        >
          <div
            class="ant-switch-handle"
          />
          <span
            class="ant-switch-inner"
          >
            <span
              class="ant-switch-inner-checked"
            />
            <span
              class="ant-switch-inner-unchecked"
            />
          </span>
        </button>
        Unanswered Questions
      </div>
      <div
        class="ant-list ant-list-split css-dev-only-do-not-override-1r287do"
      >
        <div
          class="ant-spin-nested-loading css-dev-only-do-not-override-1r287do"
        >
          <div
            class="ant-spin-container"
          >
            <ul
              class="ant-list-items"
            >
              <li
                class="ant-list-item"
              >
                <div
                  class="ant-list-item-meta"
                >
                  <div
                    class="ant-list-item-meta-content"
                  >
                    <h4
                      class="ant-list-item-meta-title"
                    >
                      <a
                        href="/question/xj352vofupe1dqz9emx13r"
                      >
                        Would you rather 
                        deploy to production once every two weeks
                         or
                         
                        deploy to production once every month
                      </a>
                    </h4>
                    <div
                      class="ant-list-item-meta-description"
                    >
                      mtsamis
                    </div>
                  </div>
                </div>
              </li>
              <li
                class="ant-list-item"
              >
                <div
                  class="ant-list-item-meta"
                >
                  <div
                    class="ant-list-item-meta-content"
                  >
                    <h4
                      class="ant-list-item-meta-title"
                    >
                      <a
                        href="/question/vthrdm985a262al8qx3do"
                      >
                        Would you rather 
                        take a course on ReactJS
                         or
                         
                        take a course on unit testing with Jest
                      </a>
                    </h4>
                    <div
                      class="ant-list-item-meta-description"
                    >
                      tylermcginnis
                    </div>
                  </div>
                </div>
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <button
        aria-checked="false"
        class="ant-switch css-dev-only-do-not-override-1r287do"
        role="switch"
        style="margin-right: 20px;"
        type="button"
      >
        <div
          class="ant-switch-handle"
        />
        <span
          class="ant-switch-inner"
        >
          <span
            class="ant-switch-inner-checked"
          />
          <span
            class="ant-switch-inner-unchecked"
          />
        </span>
      </button>
      Unanswered Questions
    </div>
    <div
      class="ant-list ant-list-split css-dev-only-do-not-override-1r287do"
    >
      <div
        class="ant-spin-nested-loading css-dev-only-do-not-override-1r287do"
      >
        <div
          class="ant-spin-container"
        >
          <ul
            class="ant-list-items"
          >
            <li
              class="ant-list-item"
            >
              <div
                class="ant-list-item-meta"
              >
                <div
                  class="ant-list-item-meta-content"
                >
                  <h4
                    class="ant-list-item-meta-title"
                  >
                    <a
                      href="/question/xj352vofupe1dqz9emx13r"
                    >
                      Would you rather 
                      deploy to production once every two weeks
                       or
                       
                      deploy to production once every month
                    </a>
                  </h4>
                  <div
                    class="ant-list-item-meta-description"
                  >
                    mtsamis
                  </div>
                </div>
              </div>
            </li>
            <li
              class="ant-list-item"
            >
              <div
                class="ant-list-item-meta"
              >
                <div
                  class="ant-list-item-meta-content"
                >
                  <h4
                    class="ant-list-item-meta-title"
                  >
                    <a
                      href="/question/vthrdm985a262al8qx3do"
                    >
                      Would you rather 
                      take a course on ReactJS
                       or
                       
                      take a course on unit testing with Jest
                    </a>
                  </h4>
                  <div
                    class="ant-list-item-meta-description"
                  >
                    tylermcginnis
                  </div>
                </div>
              </div>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
